warehouse/accounts/__init__.py

```diff
@@ -54,23 +55,29 @@ def _user(request):
     if request.identity is None:
         return None
 
-    if not isinstance(request.identity, User):
+    if isinstance(request.identity, UserTokenContext):
+        # A UserTokenContext signals a user-created API token;
+        # take the underlying user.
+        return request.identity.user
+    elif isinstance(request.identity, User):
+        return request.identity
+    else:
```

Comment:
NB: This will address https://github.com/pypi/warehouse/pull/15588#issuecomment-1996019812, by ensuring that `request.user` dispatches appropriately for `UserTokenContext`.

Reply:
This is a little gross, and makes me think `request.identity` should maybe never be a `User` and always be a `Context` that might have a user. I think this is OK for now though.

Reply:
Yeah, agreed -- @facutuesca and I can look into a follow-on for that.